R/epi_slide_opt_archive.R

```diff
@@ -0,0 +1,230 @@
+    }
+    slide_nrow <- time_delta_to_n_steps(slide_t_max - slide_t_min, time_type) + 1L
+    slide_time_values <- slide_t_min + 0L:(slide_nrow - 1L) * unit_step
+    slide_inp_backrefs <- vec_match(slide_time_values, inp_snapshot$time_value)
+    # Get additional values needed from inp_snapshot + perform any NA
+    # tail-padding needed to make slider results a fixed window size rather than
+    # adaptive at tails + perform any NA gap-filling needed:
+    slide <- inp_snapshot[slide_inp_backrefs, ]
+    slide$time_value <- slide_time_values
+    if (f_from_package == "data.table") {
+      if (before == Inf) {
+        slide[, out_colnames] <-
+          f_dots_baked(slide[, in_colnames], seq_len(slide_nrow), adaptive = TRUE)
+      } else {
+        out_cols <- f_dots_baked(slide[, in_colnames], before + after + 1L)
+        if (after != 0L) {
+          # Shift an appropriate amount of NA padding from the start to the end.
+          # (This padding will later be cut off when we filter down to the
+          # original time_values.)
+          out_cols <- lapply(out_cols, function(out_col) {
+            c(out_col[(after + 1L):length(out_col)], rep(NA, after))
+          })
+        }
+        slide[, out_colnames] <- out_cols
+      }
+    } else if (f_from_package == "slider") {
+      for (col_i in seq_along(in_colnames)) {
+        slide[[out_colnames[[col_i]]]] <- f_dots_baked(slide[[in_colnames[[col_i]]]], before = before, after = after)
+      }
+    } else {
+      cli_abort("epiprocess internal error: `f_from_package` was {format_chr_deparse(f_from_package)}",
+        class = "epiprocess__epi_slide_opt_archive__f_from_package_invalid"
+      )
+    }
+    rows_should_keep <-
+      if (before == Inf) {
+        # Re-introduce time gaps:
+        !is.na(slide_inp_backrefs)
+      } else {
+        # Get back to t1-after..t2+before; times outside this range were included
+        # only so those inside would have enough context for their slide
+        # computations, but these "context" rows may contain invalid slide
+        # computation outputs:
+        vec_rep_each(c(FALSE, TRUE, FALSE), c(before, slide_nrow - before - after, after)) &
+          # Only include time_values that appeared in the input snapshot:
+          !is.na(slide_inp_backrefs)
+      }
+    out_update <- slide[rows_should_keep, ]
+    out_diff <- tbl_diff2(prev_out_snapshot, out_update, "time_value", "update")
+    out_snapshot <- tbl_patch(prev_out_snapshot, out_diff, "time_value")
+    prev_inp_snapshot <<- inp_snapshot
+    prev_out_snapshot <<- out_snapshot
+    out_diff$version <- version
+    out_diff
+  })
+  result <- list_rbind(result)
+  result
+}
+
+#' @method epi_slide_opt grouped_epi_archive
+#' @export
+epi_slide_opt.grouped_epi_archive <- function(.x, ...) {
+  assert_set_equal(
+    group_vars(.x),
+    key_colnames(.x$private$ungrouped, exclude = c("time_value", "version"))
```

Comment:
I guess we ban grouping by non-keys? That does make sense, first time I think I've actually run into it though.

Reply:
actually would this ban grouping by a subset of key_colnames? that seems like something that we would want to support

Reply:
Like with the current `epi_slide_opt.epi_df`, we require the grouping to be by the epikeys.  There are some frustrating situations to think about when grouping by other things, such as scattered location drop-in & drop-out & how we should do completion.